core/storage/btree.rs

```diff
@@ -3013,7 +3013,7 @@ impl BTreeCursor {
                     assert_eq!(left_pointer, page.get().get().id as u32);
                     // FIXME: remove this lock
                     assert!(
-                        left_pointer <= self.pager.db_header.lock().database_size,
+                        left_pointer <= self.pager.db_header()?.database_size,
```

Comment:
Hmmm I don't know about this, if somewhere we decided to evict page 1 from cache then `db_header()` might loop internally until `io` is completed. I wonder if we should follow an approach where we do something like these two options:

1. begin_read_tx: Here we will read page 1 from pager, meaning it might read from wal or not. Then we store the contents of header in `Pager` and since we are sure this transaction will only with database header snapshot we can keep it in memory without worrying about eviction.
2. We treat `db_header` as any other potentially re-entrant code that might need `io.run_once` so this will need to return `IO` in case of not being loaded yet.

Reply:
It looks like sqlite updates databaes header in its pager shared state between connections:
```c
struct BtShared {
  Pager *pPager;        /* The page cache */
  sqlite3 *db;          /* Database connection currently using this Btree */
  BtCursor *pCursor;    /* A list of all open cursors */
  MemPage *pPage1;      /* First page of the database */

```